src/metabase/notification/send.clj

```diff
@@ -187,34 +187,19 @@
   (take-notification! [this]              "Take the next notification from the queue, blocking if none available."))
 
 (deftype ^:private NotificationQueue
-         [^java.util.LinkedList                     ids-list
-          ^java.util.concurrent.ConcurrentHashMap   id->notification
-          ^java.util.concurrent.locks.ReentrantLock queue-lock
-          ^java.util.concurrent.locks.Condition     not-empty-cond]
+         [^ArrayBlockingQueue  queue
+          ^ConcurrentHashMap id->notification]
   NotificationQueueProtocol
   (put-notification! [_ notification]
     (let [id (or (:id notification) (str (random-uuid)))]
-      (.lock queue-lock)
-      (try
-        (when-not (.containsKey id->notification id)
-          (.add ids-list id))
-        (.put id->notification id notification)
-        ;; Signal that a notification is available
-        (.signal not-empty-cond)
-        (finally
-          (.unlock queue-lock)))))
+      (when-not (.putIfAbsent id->notification id notification)
```

Comment:
hmm, this is change in behavior tho, this is keep first but we want to keep last.

Reply:
Ah right, dumb mistake though can just use `put` (returns `nil` if there was nothing)